libsolidity/analysis/ImmutableValidator.cpp

```diff
@@ -104,17 +105,47 @@ bool ImmutableValidator::visit(MemberAccess const& _memberAccess)
 
 bool ImmutableValidator::visit(IfStatement const& _ifStatement)
 {
+	// store old context variables
 	bool prevInBranch = m_inBranch;
+	VariableDeclarationSet prevWrites = std::move(m_immutableStateVariablesWrittenInBranch);
 
 	_ifStatement.condition().accept(*this);
 
 	m_inBranch = true;
+	m_immutableStateVariablesWrittenInBranch.clear();
 	_ifStatement.trueStatement().accept(*this);
 
+	// find out immutable state variables that are written in then/else branches
+	auto thenWrites = std::move(m_immutableStateVariablesWrittenInBranch);
+	VariableDeclarationSet elseWrites;
 	if (auto falseStatement = _ifStatement.falseStatement())
+	{
+		m_immutableStateVariablesWrittenInBranch.clear();
 		falseStatement->accept(*this);
+		elseWrites = std::move(m_immutableStateVariablesWrittenInBranch);
+	}
 
+	// restore and update context
 	m_inBranch = prevInBranch;
+	m_immutableStateVariablesWrittenInBranch = std::move(prevWrites);
+	std::vector<ASTNode::SourceLocation const*> bad_write_locs;
+	for (auto var: thenWrites)
+	{
+		if (elseWrites.count(var))
+		{
+			m_immutableStateVariablesWrittenInBranch.emplace(var);
+			m_initializedStateVariables.emplace(var);
+		}
+		else
+		{
+			m_stateVariablesNotInitializedInBothBranches.emplace(var);
+		}
+	}
+	for (auto var: elseWrites)
+	{
+		if (!thenWrites.count(var))
+			m_stateVariablesNotInitializedInBothBranches.emplace(var);
+	}
```

Comment:
```suggestion
	for (auto var: thenWrites)
		if (elseWrites.count(var))
			m_initializedStateVariables.emplace(var);
	m_stateVariablesNotInitializedInBothBranches += (elseWrites - thenWrites) + (thenWrites - elseWrites);
```

Reply:
Done -- also added `m_initializedStateVariables.emplace(var);` in the if branch.